crates/uv-resolver/src/pubgrub/package.rs

```diff
@@ -153,6 +160,43 @@ impl PubGrubPackage {
             }))
         }
     }
+
+    /// Create a [`PubGrubPackage`] from a package name and optional extra name.
+    pub(crate) fn from_url(
+        name: PackageName,
+        extra: Option<ExtraName>,
+        mut marker: Option<MarkerTree>,
+        url: VerbatimParsedUrl,
+    ) -> Self {
+        // Remove all extra expressions from the marker, since we track extras
+        // separately. This also avoids an issue where packages added via
+        // extras end up having two distinct marker expressions, which in turn
+        // makes them two distinct packages. This results in PubGrub being
+        // unable to unify version constraints across such packages.
+        marker = marker.and_then(|m| m.simplify_extras_with(|_| true));
+        if let Some(extra) = extra {
+            Self(Arc::new(PubGrubPackageInner::Extra {
+                name,
+                extra,
+                marker,
+                url: Some(url),
+            }))
+        } else if let Some(marker) = marker {
+            Self(Arc::new(PubGrubPackageInner::Marker {
+                name,
+                marker,
+                url: Some(url),
+            }))
```

Comment:
There is admittedly some weirdness here. If we have both a marker and an extra, I'm using the `Extra` variant. The `Marker` variant is being used _only_ for bare packages with markers. In `get_dependencies`, if `Extra` has a marker field, we flatten it out to include the base package with and without markers.

Reply:
Hmmm yes that is odd. I think that was one of the reasons I ended up not going this route initially, because it seemed like we'd want an `ExtraMarker` variant too. But I guess the flattening in `get_dependencies` works around that?

I think I would find some comments about this, on the `PubGrabInner` variants, helpful.